File: 283.move-zeroes.java
/*
 * @lc app=leetcode id=283 lang=java
 *
 * [283] Move Zeroes
 */

// @lc code=start
class Solution {
    public void moveZeroes(int[] nums) {
        // i: zero; j: non zero
        int i = 0, j = 0;
        while (j < nums.length && i < nums.length) {
            while (i < nums.length && nums[i] != 0) {
                i++;
            }
            while (j < nums.length && nums[j] == 0) {
                j++;
            }

            if (j >= nums.length || i >= nums.length) {
                break;
            }

            if (j > i) {
                int temp = nums[i];
                nums[i] = nums[j];
                nums[j] = temp;
                if (i < nums.length)
                    i++;
            } else {
                if (j < nums.length)
                    j++;
            }

        }
    }

}
// @lc code=end

File: 283.move-zeroes.py
#
# @lc app=leetcode id=283 lang=python3
#
# [283] Move Zeroes
#

# @lc code=start
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # nums before ptr1 r zero-free
        ptr1=0
        n=len(nums)
        for i in range(n):
            if nums[i]!=0:
                nums[i], nums[ptr1]=  nums[ptr1],nums[i]
                ptr1+=1
        
        return
        
        
# @lc code=end

